lua/ttt2/libraries/orm.lua

```diff
@@ -0,0 +1,267 @@
+---
+-- Object Relational Model
+-- @author Histalek
+-- @module orm
+
+if SERVER then
+	AddCSLuaFile()
+end
+
+local sql = sql
+
+orm = orm or {}
+
+local cachedModels = cachedModels or {}
+
+local ORMMODEL = {}
+local ORMOBJECT = {}
+
+---
+-- Returns an object relational model according to the specified databasetable. Does nothing if no databasetable with the given name exists.
+-- @param string tableName The name of the table in the database to create a model for.
+-- @param[opt] boolean force If set to `true` the function will not return a cached version of the model.
+-- @return ORMMODEL|nil Returns the model of the database table or nil if the database table does not exist.
+-- @realm shared
+function orm.Make(tableName, force)
+	if IsValid(cachedModels[tableName]) and not force then
+		return cachedModels[tableName]
+	end
+
+	if not sql.TableExists(tableName) then return end
+
+	local primaryKey = sql.GetPrimaryKey(tableName)
+	local dataStructure = sql.GetTableColumns(tableName)
+
+	if not primaryKey or not dataStructure then
+		ErrorNoHalt("[ORM] An sql error occurred while retrieving the metadata of the following databasetable: " .. tableName)
+	end
+
+	local model = {}
+	model._tableName = tableName
+	model._dataStructure = dataStructure
+	model.All = ORMMODEL.All
+	model.New = ORMMODEL.New
+	model.Where = ORMMODEL.Where
+
+	if #primaryKey == 0 then
+		return model
+	end
+
+	model._primaryKey = primaryKey
+	model.Find = ORMMODEL.Find
+
+	-- Prepare strings that will not change unless the model itself changes. So we don't have to create these strings everytime we use `model.Save()`.
+	local columnList = {nil, nil}
+
+	for i = 1, #primaryKey do
+		columnList[i] = sql.SQLIdent(primaryKey[i])
+	end
+
+	model._primaryKeyList = table.concat(columnList, ",")
+
+	for i = 1, #dataStructure do
+		columnList[i] = sql.SQLIdent(dataStructure[i])
+	end
+
+	model._columnList = table.concat(columnList, ",")
+
+	cachedModels[tableName] = model
+
+	return model
+end
+
+---
+-- @class ORMMODEL
+
+---
+-- Retrieves all saved objects of the model from the database.
+-- @return table|nil Returns an array of all found objects or nil in case of an error.
+-- @realm shared
+function ORMMODEL:All()
+	local objects = sql.Query("SELECT * FROM " .. sql.SQLIdent(self._tableName))
+
+	if objects == false then return end
+
+	if not self._primaryKey then
+		return objects
+	end
+
+	for i = 1, #objects do
+		objects[i].Save = self.Save
+		objects[i].Delete = self.Delete
+		objects[i].Refresh = self.Refresh
+	end
+
+	return objects or {}
+end
+
+---
+-- Retrieves a specific object by their primarykey from the database.
+-- @param number|string|table primaryValue The value(s) of the primarykey to search for.
```

Comment:
probably `any` would match better (if booleans are supported as well, they aren't listed here)

Reply:
actually only strings and a table of strings are supported ^^ will fix